Landing page: Add multiple communities list and modals

jennur opened the pull request:
* closes https://github.com/inveniosoftware/invenio-app-rdm/issues/2061
* closes https://github.com/inveniosoftware/invenio-stats/issues/134

## Screenshots
<img width="599" alt="Screenshot 2023-03-06 at 10 57 44" src="https://user-images.githubusercontent.com/38131488/223077467-54ae9bcf-5e48-4137-b8cb-0da706477da6.png">


### Loading state
<img width="373" alt="Screenshot 2023-02-23 at 16 56 06" src="https://user-images.githubusercontent.com/21052053/220961430-54cce3d5-740c-4d10-b338-9c16777a39fc.png">


### Error state
<img width="372" alt="Screenshot 2023-02-23 at 16 55 43" src="https://user-images.githubusercontent.com/21052053/220961399-63fd3607-39cf-4d72-b811-9df72d52f916.png">


### "View all" modal
<img width="1675" alt="Screenshot 2023-03-06 at 10 46 28" src="https://user-images.githubusercontent.com/38131488/223077374-21415761-8578-4c01-9c09-59dfbd16b796.png">

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesList.js:
```diff
@@ -0,0 +1,116 @@
+  }
+}
+
+// CommunitiesList.propTypes = {
```
leftovers?

kpsherva replied:
I think it is OK already to have them, no need to comment out. Instead of passing recordId as a prop, and then hardcoding the URL as `/api/records/{recordDd}/communities` (which is not ideal solution) you could simply pass the whole endpoint `recordCommunitySearchEndpoint` as a prop and set it in jinja. It will be more flexible to change in the future if done this way

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesList.js:
```diff
@@ -0,0 +1,116 @@
+
+import { i18next } from "@translations/invenio_app_rdm/i18next";
+import React, { Component } from "react";
+// import PropTypes from "prop-types";
```
👀

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesList.js:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  componentWillUnmount() {
+    this.cancellableFetchCommunities?.cancel();
```
instead of using `?` operator maybe it makes sense to declare it as a class attribute?

jennur replied:
I'm not sure what you mean? It's already defined in here https://github.com/inveniosoftware/react-invenio-forms/blob/master/src/lib/elements/withCancel.js

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesList.js:
```diff
@@ -0,0 +1,116 @@
+    try {
+      const res = await this.cancellableFetchCommunities.promise;
+      this.setState({
+        communities: res.data?.hits?.hits,
```
maybe destructuring assignment would work better ?

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesList.js:
```diff
@@ -0,0 +1,116 @@
+
+  fetchRecordCommunities = async () => {
+    // TODO: Update endpoint to be record specific
+    return await axios.get("/api/communities", {
```
temporarily, you could do (to already have the correct communities for testing):
```suggestion
    return await axios.get("/api/communities?q=id:${community.id} OR id:${community.id} OR ...", {
```
where `q=id:${community.id} OR id:${community.id}` this part could be generated base on communities included in the record

jennur replied:
But it is not the record id we want to include? I.e. something like `/api/record/{id}/communities` ?

jennur replied:
If it's only for testing, is it necessary to change?

kpsherva replied:
> But it is not the record id we want to include in the end? I.e. something like `/api/records/{id}/communities` ?

yes this is how it will look like at the end, but until then you could search in communities endpoint. Either way it was just a suggestion how to already display correct data before the endpoint is completed

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesList.js:
```diff
@@ -0,0 +1,116 @@
+  render() {
+    const { communities, loading, error } = this.state;
+    const communityItems = communities?.map((community, index) => {
+      if (index > 2) return null;
```
if you consider the query from the previous comment, this line could be probably removed

jennur replied:
Yes

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesListModal/CommunitiesListModal.js:
```diff
@@ -0,0 +1,63 @@
+        open={modalOpen}
+        onClose={() => this.setState({ modalOpen: false })}
+        onOpen={() => this.setState({ modalOpen: true })}
+        trigger={React.cloneElement(trigger, {
```
just out of curiosity: why did you have to use the `cloneElement` for the trigger?

jennur replied:
You're right, it's not needed! I just copied the modal from somewhere else where this was implemented, and wasn't thinking :D

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesListModal/CommunityItem.js:
```diff
@@ -0,0 +1,57 @@
+          {result.metadata.title}
+        </a>
+
+        {result.access.visibility === "restricted" && (
```
maybe we could make it a common component in communities ? I think we use it many times around so we wouldn't have to copy the code

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesListModal/CommunityItem.js:
```diff
@@ -0,0 +1,57 @@
+      </Item.Header>
+
+      <Item.Description
+        content={_truncate(result.metadata.description, { length: 50 })}
```
discussed in person: needs html escaping

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesListModal/RecordCommunitiesSearch.js:
```diff
@@ -0,0 +1,76 @@
+    // TODO: Update with record specific endpoint
+    const searchApi = new InvenioSearchApi({
+      axios: {
+        url: "/api/communities",
```
same as above

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/index.js:
```diff
@@ -67,3 +70,19 @@ if (recordExportDownloadDiv) {
 }
+
+const submitToCommunityModalDiv = document.getElementById("submit-to-community-modal");
+ReactDOM.render(
```
discussed in person: this part need to be refactored as a part of another task. We cannot simply import it from react-invenio-deposit becuase deposit components should not be a part of the landing page

jennur commented:
## Updates

- Extracted the restricted label as a separate compontent
- Removed the submit-to-community modal (will be implemented in a separate task
- Removed empty constructor
- Fixed HTML escaping
- Removed `cloneElement` from trigger

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesListModal/CommunityItem.js:
```diff
@@ -0,0 +1,55 @@
+    </Item.Content>
+
+    <div className="flex align-items-center">
+      <Button as="a" size="tiny" href={`/communities/${result.slug}`}>
```
no need to hardcode this url, it is available in the community object under `links`

kpsherva commented on invenio_app_rdm/theme/assets/semantic-ui/js/invenio_app_rdm/landing_page/CommunitiesListModal/CommunityItem.js:
```diff
@@ -0,0 +1,55 @@
+
+    <Item.Content verticalAlign="middle">
+      <Item.Header as="h3" className="ui small header flex align-items-center mb-5">
+        <a href={`/communities/${result.slug}`} className="p-0">
```
same as below

jennur commented:
- Removed "View community" button
![Screenshot 2023-02-24 at 10 39 38](https://user-images.githubusercontent.com/21052053/221145264-d75ff6a7-d7e5-4243-93c4-0033d5481548.png)

jennur commented:
- Changed color of view-all-button
- Added number of communities in the button text

![Screenshot 2023-02-24 at 10 54 01](https://user-images.githubusercontent.com/21052053/221148793-5ec10c4a-ff6e-4f02-897f-17c1803022b8.png)

jennur commented:
<img src="https://user-images.githubusercontent.com/38131488/223077374-21415761-8578-4c01-9c09-59dfbd16b796.png" />

- Should the labels not be above the title as we have in the other list items (e.g. in record search)?
- Is the image placeholder broken?
- It's not super clear what the chat-icon is for. Should they not only be for pending communities? It also looks a bit floating in the middle. In the mockups it's positioned below the action buttons ("View", "Cancel" - in pending communities only).

ntarocco commented on invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/communities.html:
```diff
@@ -0,0 +1,27 @@
+  {# Communities list #}
+  {#  TODO: Update endpoint to {{ record.links["communities"] }} specific #}
+  {%  set ids = "id:" + record.parent.communities.ids | join (' OR id:')  %}
+  <div id="record-communities-list" data-record-community-search-endpoint="/api/communities?q={{ ids }}"></div>
```
I think that we should not use the `q`, because if in the future we add a search bar, then the query will be displayed and the user might be able to change them.
There was this `hidden` field at some point or we can find better solution.

kpsherva replied:
this will be replaced with result.links["communities"], it is temporary

ntarocco commented on invenio_app_rdm/communities_ui/templates/semantic-ui/invenio_communities/details/index.html:
```diff
@@ -13,7 +13,7 @@
 
 {%- block javascript %}
     {{ super() }}
-    {{ webpack['invenio-communities-details-search.js'] }}
```
should this bundle be removed from where it was created then?

ntarocco commented on invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/stats.html:
```diff
@@ -25,138 +25,85 @@
-  <p class="m-auto rel-mt-1 rel-mb-1 text-muted">
-    <i>{{ _("All statistics are updated daily.")}} <a href="/help/statistics">{{ _("Read more") }}...</a></i>
+  <p class="text-muted">
+    <a href="/help/statistics">{{ _("More info on how stats are collected.") }}...</a>
```
Without the dot?
```suggestion
    <a href="/help/statistics">{{ _("More info on how stats are collected") }}...</a>
```
Will the static page `/help/statistics` always be there?

kpsherva replied:
yes, it is hardcoded for now

ntarocco commented on invenio_app_rdm/records_ui/templates/semantic-ui/invenio_app_rdm/records/details/side_bar/metrics.html:
```diff
@@ -8,8 +8,9 @@
 #}
 
-
+{%  if stats %}
```
This `if` looks a bit ugly, is it needed?

kpsherva replied:
yes, it fixes broken context

ntarocco commented on invenio_app_rdm/theme/assets/semantic-ui/less/invenio_app_rdm/theme/elements/image.overrides:
```diff
@@ -52,7 +52,7 @@
 
   img {
-    height: @tinyWidth !important; // needs to override height set for images in items by SUI
+    height: 115px !important; // needs to override height set for images in items by SUI
```
as chat, we can extract this in a var?
